### PR TITLE
meson: use template to declare executables and modules

### PR DESCRIPTION
**Pull request by yuwata**
split-out of #28042.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -2478,6 +2569,61 @@ foreach tuple : [['myhostname', 'ENABLE_NSS_MYHOSTNAME'],
...
+                kwargs += { key : val }
+        endforeach
+
+        sym = meson.current_source_dir() / '@0@'.format(dict.get('link_depends')[0])
+
+        # Note that we link NSS modules with '-z nodelete' so that mempools never get orphaned
+        link_args = ['-shared',
+                     '-Wl,--version-script=' + sym]
+        if is_nss
+                link_args += ['-Wl,-z,nodelete']
```
Comment (medhefgo) [inline meson.build]: Not that matters, but `-Wl,` should not be needed as the compiler understands `-z`...

File: meson.build [hunk 2]
```diff
@@ -2478,6 +2569,61 @@ foreach tuple : [['myhostname', 'ENABLE_NSS_MYHOSTNAME'],
...
+        endif
+
+        lib = shared_library(
+                name,
+                link_args : link_args,
```
Comment (medhefgo) [inline meson.build]: How does this interact if `link_args` is passed in the dict/kwargs?

File: meson.build [hunk 3]
```diff
@@ -2170,6 +2170,9 @@ simple_tests = []
 fuzzers = []
 simple_fuzzers = []
 catalogs = []
+modules = [] # nss, pam, and other plugins
```
Comment (medhefgo) [inline meson.build:2177]: Since it creates shared libraries, shouldn't this be named as such?

**Review by medhefgo — APPROVED** [review]
Man, meson really ought to have local variables. The amount of variables we use makes me fear we accidentally shadow/override something somewhere.

File: meson.build (hunk 2 above)
Comment (yuwata) [reply meson.build]: Such conflict is handled as critical, and meson refuses to build.

File: meson.build (hunk 3 above)
Comment (yuwata) [reply meson.build:2177]: Currently, `libsystemd-shared`, `libsystemd`, or `libudev` are not built with this way... So, I'd like not to name the array shared_libraries or so.

**Comment by yuwata** [comment]
@medhefgo Thank you for the review. Addressed the suggestion. Setting the green label.

**Comment by bluca** [comment]
> @medhefgo Thank you for the review. Addressed the suggestion. Setting the green label.

I think we should wait for @keszybz to have a look too

**Comment by yuwata** [comment]
Let's postpone this, and merge after v254 is released.

File: meson.build [hunk 4]
```diff
@@ -2438,6 +2378,7 @@ subdir('src/busctl')
 subdir('src/coredump')
 subdir('src/cryptenroll')
 subdir('src/cryptsetup')
+subdir('src/cryptsetup/cryptsetup-tokens')
```
Comment (keszybz) [inline meson.build]: Not that it matters greatly, but it'd be more usual to do move this include to `src/cryptsetup/meson.build`.

File: src/cryptsetup/cryptsetup-tokens/meson.build [hunk 5]
```diff
@@ -1,7 +1,8 @@
 # SPDX-License-Identifier: LGPL-2.1-or-later
 
-cryptsetup_token_sym = files('cryptsetup-token.sym')
-cryptsetup_token_sym_path = meson.current_source_dir() / 'cryptsetup-token.sym'
+if conf.get('HAVE_LIBCRYPTSETUP_PLUGINS') != 1
+        subdir_done()
+endif
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup-tokens/meson.build]: Why not just put `conditions : ['HAVE_LIBCRYPTSETUP_PLUGINS']` in the definition?

File: src/nss-mymachines/meson.build [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+if conf.get('ENABLE_NSS_MYMACHINES') != 1
+        subdir_done()
+endif
```
Comment (keszybz) [inline src/nss-mymachines/meson.build]: … and here too, why not `conditions : ['ENABLE_NSS_MYMACHINES']` in the definition? That way the rest of the file is "executed" even if something is skipped in the local configuration, so it's less likely to break in other configurations.

**Review by keszybz — COMMENTED** [review]
Looks reasonable. I guess if we convert all binaries, we'll have some savings in the number of lines.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. Addressed the comments. PTAL.

**Pull request merged**